hmmlearn/hmm.py

```diff
@@ -520,134 +520,221 @@ class GMMHMM(_BaseHMM):
-        super(GMMHMM, self)._init(X, lengths=lengths)
+        super(GMMHMM, self)._init(X, lengths)
 
-        for g in self.gmms_:
-            g.set_params(init_params=self.init_params, n_iter=0)
-            g.fit(X)
+        _, n_features = X.shape
+        self.n_features = n_features
 
-    def _compute_log_likelihood(self, X):
-        return np.array([g.score(X) for g in self.gmms_]).T
+        main_kmeans = cluster.KMeans(n_clusters=self.n_components,
+                                     random_state=self.random_state)
+        main_kmeans.fit(X)
+        labels = main_kmeans.labels_
+        kmeanses = []
+        for i, label in enumerate(np.unique(labels)):
+            kmeans = cluster.KMeans(n_clusters=self.n_mix,
+                                    random_state=self.random_state)
+            kmeans.fit(X[np.where(labels == label)])
+            kmeanses.append(kmeans)
+
+        if 'w' in self.init_params or not hasattr(self, 'weights_'):
+            self.weights_ = (np.ones((self.n_components, self.n_mix)) /
+                             (np.ones((self.n_components, 1)) * self.n_mix))
+
+        if 'm' in self.init_params or not hasattr(self, 'means_'):
+            self.means_ = np.zeros((self.n_components, self.n_mix,
+                                    self.n_features))
+            for i, kmeans in enumerate(kmeanses):
+                self.means_[i] = kmeans.cluster_centers_
+
+        if 'c' in self.init_params or not hasattr(self, 'covars_'):
+            cv = np.cov(X.T)
+            if not cv.shape:
+                cv.shape = (1, 1)
+
+            if self.covariance_type == 'tied':
+                self.covars_ = np.zeros((self.n_components,
+                                         self.n_features, self.n_features))
+                self.covars_[:] = cv
+            elif self.covariance_type == 'full':
+                self.covars_ = np.zeros((self.n_components, self.n_mix,
+                                         self.n_features, self.n_features))
+                self.covars_[:] = cv
+            elif self.covariance_type == 'diag':
+                self.covars_ = np.zeros((self.n_components, self.n_mix,
+                                         self.n_features))
+                self.covars_[:] = np.diag(cv)
+            elif self.covariance_type == 'spherical':
+                self.covars_ = np.zeros((self.n_components, self.n_mix))
+                self.covars_[:] = cv.mean()
+
+    def _check(self):
+        super(GMMHMM, self)._check()
 
     def _generate_sample_from_state(self, state, random_state=None):
-        return self.gmms_[state].sample(1, random_state=random_state).flatten()
+        # TODO: actually use random_state
+        if random_state is None:
+            random_state = self.random_state
+
+        cur_means = self.means_[state]
+        cur_covs = self.covars_[state]
+        cur_weights = self.weights_[state]
+
+        i_gauss = np.random.multinomial(1, cur_weights)[0][0]
```

Comment:
You want Categorical distribution here. Try [`np.random.choice`](http://docs.scipy.org/doc/numpy-1.10.0/reference/generated/numpy.random.choice.html) with `p` argument.